README.md
For more blockchain books, see the [**Best of Crypto Books page »**](https://openblockchains.github.io/crypto-books/)

---


![](i/cryptokitties-genes01.png)


The Future of Digital Collectibles - CryptoKitties, CryptoPuppies, CryptoPets, CryptoMonsters, CryptoTulips - Latest (and Greatest) Investment Opportunities in Online Casino Gambling - The Future is Meow

> - Fabulous Persian Spock Gerbil Gold Cottoncandy - Extremely rare gen 5 swift virgin | 2.9 ETH
> - Rarity: 0.00264% Gen 5 JAGUAR FABULOUS GOLD DALI!! VIRGIN!
> - Rarity: 0.0015% Princess Bubblegum is now for sale! Gen 12 | Brisk | Virgin | Chartreux | Bubblegum | Otaku | Emeraldgreen | Saycheese | Mauveover | Spock - Starts ETH 20/Ends ETH 10
> - Gold ducat, Gen 5, Virgin, Swift. Very cheap
> - Cheap Gen 1 cute kittie with rare genes! Only 0.125 ETH
> - UNIQUE Virgin Peach Googly Gold Mauveover gen:2 cooldown:1 0.87992% RARE
> - SUPER CHEAP: Gerbil, Ragdoll, Scarlet, Chestnut, Cotton Candy!!! 0.02 ETH (~$14)
> - I'm giving away a Gen 1 FAST Gold for free...
>
> -- [CrypotKittiesMarket](https://www.reddit.com/r/CryptoKittiesMarket)


> Blockchain has unlocked the magic of digital scarcity, and combining that with the power of
> making the digital goods persistent gives them a potential value that is only limited by how much
> prestige a wealthy person might place on ownership of the item.
>
> -- [Justin Poirier](https://twitter.com/tokenizedcap/status/938460753589424128)


> 3 days of up and 19 days of down down down... And the 3 up days were right at the start. Yeah, looking real good.
>
> Nobody wants to pay $300 for a hexadecimal string. That's LITERALLY all you’re buying.
> All the website does is display a picture of a cat. You're paying hundreds of dollars for a picture of a cat.
> Let that sink in.
>
> This was a modern tulip bubble. You were buying hex strings that tied to the picture of a cat.
> People were paying 10-20-30 ETH for a cat picture. 10 ETH was around $4000? For a cat picture. Yeah...
>
> -- [CryptoKitties](https://www.reddit.com/r/CryptoKitties)



**REMINDER: In the digitial world there are no originals! Every copy is a original and you cannot tell the difference
(all 0s and 1s are the same). And, yes, you can always make as many (free) copies as you like (in a free world).
Claiming that you can protect your exclusive rights to pixels because the record of ownership
is stored in a public database is absurd.
Without laws and governments that help you with your rights - the claim is just
a meaningless series of 0s and 1s.  PS: Do you really own CryptoKitties? (Spoiler: No.)
Check your license agreement with DapperLabs - the pixel art license seller - and do NOT get fooled by the record of ownership.
You are a licensee and NOT an owner.**


From the [CryptoKitties License F.A.Q](https://www.niftylicense.org):

> Dapper Labs, Inc. created [and owns] the art of your cryptokitty, and it's a big part of our brand.
> At the end of the day, businesses also need to have the right to make the big decisions [and the big $$$]
> about how the art gets used in a commercial setting.
>
> We reserve to ourselves the right to make bigger [and bigger] deals to help our business grow
> [and rake in all the $$$ for ourselves - of course].




# Awesome CryptoKitties Bubble (Anno 2017)

A collection about Awesome CryptoKitties (Yes, Cute Little Cartoon Cats) on the Blockchain! and CryptoCopycats - digital collectibles secured on a distributed public databases w/ crypto hashes.  Are CryptoPuppies the new CryptoKitties?  Contributions welcome.


What's News?

For the latest updates, see [**CryptoKitties in 2021 & 2020 »**](NEWS.md)


Contents

- [CryptoKitties (Yes, Cute Little Cartoon Cats) on the Blockchain!](#cryptokitties-yes-cute-little-cartoon-cats-on-the-blockchain)
  - [Press Coverage](#press-coverage) - _CryptoKitties Mania (Anno 2017)_
  - [Timeline](#timeline) - _Major Milestones in the CryptoKitties History_
  - [Owners / Producers](#owners--producers)
    - [Team](#team) - _Meet the CryptoKitties Makers_
  - [Special Collector's CryptoKitties](#special-collectors-cryptokitties)
  - [Sales Statistics $$$](#sales-statistics)
  - [Articles](#articles)
  - [Money, Money, Money $$$](#money-money-money-) - _Gambling with CryptoKitties - Welcome to the Casino - Get-Rich-Quick_
    - [How Do the Founder Make Money with CryptoKitties?](#how-do-the-founder-make-money-with-cryptokitties)
    - [How Can You Make Money with CryptoKitties?](#how-can-you-make-money-with-cryptokitties)
    - [Scams, Scams, Scams - The Art of the Steal](#scams-scams-scams---the-art-of-the-steal) - _Buyer Beware! Seller Beware! Here Be Dragons!_
    - [Is CryptoKitties a Scam? Is CryptoKitties a Pyramid (Gambling) Scheme? Is CryptoKittes an Online (Gambling) Casino?](#is-cryptokitties-a-scam-is-cryptokitties-a-pyramid-gambling-scheme-is-cryptokittes-an-online-gambling-casino)
      - [Q: Is CryptoKitties a scam?](#q-is-cryptokitties-a-scam)
      - [Q: Are CryptoKitties like dollars or like stuffed animals?](#q-are-cryptokitties-like-dollars-or-like-stuffed-animals)
      - [Q: Is CryptoKitties a pyramid (gambling) scheme?](#q-is-cryptokitties-a-pyramid-gambling-scheme)
      - [Q: Is CryptoKitties an online gambling casino?](#q-is-cryptokitties-an-online-gambling-casino)
  - [Fun, Fun, Fun](#fun-fun-fun) - _Modern Art Paintings (Acryl on Canvas) and More_
  - [Questions & Answers](#questions--answers)
    - [Tech](#tech)
       - [Q: Couldn't CryptoKitties be hosted on many types of databases besides a blockchain?](q-couldnt-cryptokitties-be-hosted-on-many-types-of-databases-besides-a-blockchain)
    - [Legal](#legal)
       - [Q: Do you really own your CryptoKitties?](#q-do-you-really-own-your-cryptokitties) - _Spoiler: No_
    - [Casino & Gambling / "Market & Investing"](#casino--gambling--market--investing)
      - [Q: What's the cost of playing?](#q-whats-the-cost-of-playing)
      - [Q: What's the population? How many gen0 kitties will there be?](#q-whats-the-population-how-many-gen0-kitties-will-there-be)
      - [Q: What's a clock auction?](#q-whats-a-clock-auction)
      - [Q: What make's a kitty good?](#q-what-makes-a-kitty-good)
      - [Q: What are rare traits?](#q-what-are-rare-traits)
  - [Inside CryptoKitties Genetics](#inside-cryptokitties-genetics)
  - [Code on the Blockchain - Electronic Contract Scripts](#code-on-the-blockchain---electronic-contract-scripts)
  - [More / Misc](#more--misc)
- [Crypto Copycats on the Blockchain!](#crypto-copycats-on-the-blockchain)
- [History Corner](#history-corner) - _Collectibles on the Blockchain before CryptoKitties_
  - [CryptoPunks on the Blockchain!](#cryptopunks-on-the-blockchain)




## CryptoKitties (Yes, Cute Little Cartoon Cats) on the Blockchain!

Collectible. Breedable. Adorable.

Collect and breed digital cats. Start meow. Buy! Sell! Hold!

> Q: What's CryptoKitties? What's the big deal?
>
> CryptoKitties is centered around breedable, collectible, and oh-so-adorable creatures
> we call CryptoKitties! Each cat is one-of-a-kind and 100% owned by you (*);
> it cannot be replicated, taken away, or destroyed.
>
> CryptoKitties is built on blockchain technology.
> You can buy, sell, or trade your CryptoKitty like it was a traditional collectible,
> secure in the knowledge that blockchain will track ownership securely.
> But, unlike traditional collectibles, you can breed two CryptoKitties
> to create a brand-new, genetically unique offspring.
> It results in something special—just like you!

(*): See Q: Do you really own your CryptoKitties? A: Spoiler: No.


Learn more @ [cryptokitties.co](https://cryptokitties.co),
medium: [cryptokitties](https://medium.com/cryptokitties),
twitter: [CryptoKitties](https://twitter.com/CryptoKitties),
reddit: [CryptoKitties](https://www.reddit.com/r/CryptoKitties)


![](i/cryptokitties.png)


> All I want for Christmas is a CryptoKitty.
>
> -- [Kayla Williams](https://twitter.com/kaylaw/status/938590748655550464)

> I got a fever. And the only prescription is more CryptoKitties.
>
> -- [Eduardo Salazar](https://twitter.com/ceduardosalazar/status/938558630663634944)

> My Gen 7 CryptoKitty. The Future is Meow.
>
> -- [Anshul Dhawan](https://twitter.com/TheAnshulDhawan/status/938551642202324993)

> Celebrating 100 000 Kitties!
>
> -- [CryptoKitties](https://twitter.com/CryptoKitties/status/938223161232916481) - Dec/5

> WikiLeaks now accepts #CryptoKitties as a contribution method.
> In many jurisdictions you can write them down for tax purposes as an investment.
> [Official WikiLeaks CryptoKitties](https://wikileaks.shop/pages/cryptokitties) -
> Purebred WikiLeaks CryptoKitties arrive in time for Christmas.
>
> -- [WikiLeaks](https://twitter.com/wikileaks/status/944209405377101824)



### Press Coverage

- [**People have spent over $1M buying virtual cats on the Ethereum blockchain**](https://techcrunch.com/2017/12/03/people-have-spent-over-1m-buying-virtual-cats-on-the-ethereum-blockchain/), TechCrunch. Dec/3
- [**CryptoKitties Mania Overwhelms Ethereum Network's Processing**](https://www.bloomberg.com/news/articles/2017-12-04/cryptokitties-quickly-becomes-most-widely-used-ethereum-app), Bloomberg, Dec/4
- [**Introducing CryptoKitties, the New Digital Pets Taking Ethereum by Storm**](http://fortune.com/2017/12/04/blockchain-cryptokitties-ethereum/), Fortune, Dec/4
- [**This Game Combines The Internet's Favorite Things: Cats & Cryptocurrency**](https://www.youtube.com/watch?v=qk7gRljIKww), Vice, Dec/5
- [**Meet CryptoKitties, the $100,000 digital beanie babies epitomizing the cryptocurrency mania**](https://www.cnbc.com/2017/12/06/meet-cryptokitties-the-new-digital-beanie-babies-selling-for-100k.html), CNBC, Dec/6
- [**Ethereum Network Copes With Surge of Activity as Virtual Kitten Game Goes Viral**](https://blogs.wsj.com/cio/2017/12/07/ethereum-network-copes-with-surge-of-activity-as-virtual-kitten-game-goes-viral/), The Wall Street Journal, Dec/7
- [**CryptoKitties have the digital currency world purring**](http://www.sfchronicle.com/news/article/CryptoKitties-have-the-digital-currency-world-12462575.php),
The San Francisco Chronicle,
Dec/27 - The digital kitties might not be worth the paper they're not printed on.
- [**Reigning Cats -How to buy cryptocurrency Ethereum – is it worth as much as Bitcoin and what are CryptoKitties?**](https://www.thesun.co.uk/money/5087162/how-to-buy-ethereum-price-cryptocurrency-cryptokitties-bitcoin/), The Sun, Dec/28 - Players can buy and sell virtual cats for as much as $10,000 on the new digital currency - with their rather valuable 'Cattributes'
- [**CryptoKitties, Explained ... Mostly**](https://www.nytimes.com/2017/12/28/style/cryptokitties-want-a-blockchain-snuggle.html), The New York Times, Dec/28 - Why are $20 million and 180,000 people suddenly in the market for digital cats? We gamified the blockchain.



### Timeline

Note: For official (twice weekly) updates / timeline see [updates.cryptokitties.co](https://updates.cryptokitties.co).


_CryptoKitties History_

- 2017 Oct/19 - CryptoKitties alpha test starts at ETH Waterloo (in Ontario, Canada) - an Ethereum hackathon; hundreds of users bred thousands of cats
- 2017 Nov/22 - CryptoKitties contract scripts uploaded onto the public Ethereum blockchain
- 2017 Nov/23 - CryptoKitties closed beta starts; one of the top three most active Ethereum contracts within three hours
- 2017 Nov/28 - CryptoKitties it's official; open to the public - let's wecome everyon to collect, buy, sell, and breed digital cartoon cats on the blockchain
- 2017 Nov/28 - $4_700 USD kitty sold on day of launch
- 2017 Nov/30 - 873 users with an average of 15 kitties each; 19_264 CryptoKitties created or bred; over 120 ETH (~$51_000) in transcations within 1 week; 3.5% of ALL Ethereum transactions (2nd contract on the blockchain)  <!-- source: https://twitter.com/CryptoKitties/status/936349761644273664 -->
- 2017 Dec/2  - Highest selling cat - Kitty #1 (Gen 0, Genesis Exclusive) - sold for Ξ 246.9255 ETH (~$114_481.59)
- 2017 Dec/5  - ~60k registered users; over $5M USD in transactions; CryptoKitties accounts for ~25% of world-wide Ethereum traffic
- 2017 Dec/6  - Celebrating 100 000 CryptoKitties  <!-- check date? add tweet -->
- 2017 Dec/6  - Highest selling cat - Kitty #4 (Gen 0) - sold for Ξ 247.0000 ETH (~$107_816.49)
- 2017 Dec/7  - Highest selling cat - Kitty #18 (Gen 0) - sold for Ξ 253.3368 ETH (~$110_707.16)
- 2017 Dec/10 - Celebrating 200 000 CryptoKitties
- 2017 Dec/12 - 150k+ registered users; $15M USD in transactions
- 2017 Dec/14 - Celebrating 300 000 CryptoKitties
- 2017 Dec/21  - WikiLeaks Accepts (Sells) CryptoKittes for Donations  <!-- check date? add tweet  -->
- 2017 Dec/24 - Celebrating 400 000 CryptoKitties

<!-- new year -->

- 2018 Feb/14 - Family (breeding "bloodline") jewels (for "mewtations"/mutations) added / live
- 2018 Feb/19 - Official blog launched
- 2018 Feb/23 - Fancy Cats now limited edition
- 2018 Mar/15 - Enhanced cattributes added / live



_From the Official CryptoKitties News Blogs_

- 2018 Feb/19 - [How to get started and stay up to date on CryptoKitties](https://medium.com/cryptokitties/how-to-get-started-and-stay-up-to-date-on-cryptokitties-15e2a5fb8f17)

> Moving forward, we're going to treat this blog as our source of truth [...]
> this blog will be how you get the latest in-depth news for everything CryptoKitties.

- 2018 Feb/23 - [All Fancy Cats are now limited edition!](https://medium.com/cryptokitties/all-fancy-cats-are-now-limited-edition-717dfa8e20e)
- 2018 March/12 - [All about Family Jewels](https://medium.com/cryptokitties/all-about-family-jewels-ee4bc1f56608) -- A guide to Family Jewels, mewtations, and how they all work together
- 2018 March/20 - [Cat's out of the bag - CryptoKitties receives $12M in funding](https://medium.com/cryptokitties/cryptokitties-receives-12m-in-funding-38322f90ab65)

> Who are these fat cats investing in CryptoKitties? The investment is led by Andreessen Horowitz
> and Union Square Ventures, two of the world's top venture capital firms...


Big in Greater China? Updates

- 2018 March/7 - [Big news on CryptoKitties iOS App launch in China, Hong Kong, and Taiwan!](https://medium.com/cryptokitties/big-news-on-cryptokitties-ios-app-launch-in-china-hong-kong-and-taiwan-bf767719fe22) -- Giveaways, new cats, our first Artist Series contributor will precede the launch
- 2018 March/8 - [CryptoKitties: Now Made in China™](https://medium.com/cryptokitties/cryptokitties-now-made-in-china-b8eddfaed599)
- 2018 March/21 - [So how's it going in China?](https://medium.com/cryptokitties/so-hows-it-going-in-china-6b5c512b58f4)




_Upcoming (Planed) Milestones_

- Spring 2018  - The CryptoKitties mobile app starts
- Early 2018   - CryptoKitties plans to accept dollars, euros and yen or credit cards as soon as possible.
- November 2018 - The last of the 50_000 Gen 0 CryptoKitties auctioned off to the public; CryptoKitties can now only grow by breeding.


_More (Minor) Milestones_

<!-- break
     move to traits? -->

- 2017 Dec/3 -  Due to network congestion, we are increasing the birthing fee from 0.001 ETH to 0.002 ETH. This will ensure your kittens are born on time! The extra is needed to incentivize miners to add birthing txs to the chain. Long-term solution will be explored very soon!
- 2017 Dec/4 - DUE TO CONGESTION ON THE NETWORK WE ARE INCREASING THE BIRTHING FEE FROM 0.002 ETH TO 0.015 ETH. WE HOPE THIS WILL ALLOW MORE KITTENS TO BE BORN ON TIME AS THERE HAVE BEEN SIGNIFICANT DELAYS
- 2017 Dec/6 - WE ARE INCREASING THE RECOMMENDED DEFAULT GAS PRICE FROM 25 TO 40 GWEI IN ORDER TO INCENTIVIZE THE MINERS. AS ALWAYS YOU ARE ABLE TO ADJUST THIS AT YOUR OWN DISCRETION
- 2017 Dec/11 - We have dropped the birthing fee from 0.015 ETH to 0.008 ETH. Happy breeding everyone!
- 2017 Dec/22-24 - Winter CryptoKitties Giveaway Contest
- 2018 March/13 - [Why does our database fall out of sync?](https://medium.com/cryptokitties/why-does-our-database-fall-out-of-sync-c86b1eefa6d)


_New Traits / Fancies Timeline_

2017

- Dec/20  - New Trait - Serpent - Eye type
- Dec/20  - New Trait - Alien - Eye type
- Dec/22  - New Trait - Hintomint  - Body color
- Dec/28  - New Trait - Elk - Discovered in Kitty #422103.
- Dec/31  - New Trait - Icy  - Secondary color (chest color) - white
- Dec/31  - New Trait - Selkirk - Body type - hair puff with thin tail
- Dec/31  - New Trait - Flamingo - Secondary color (chest color) - pink
- Dec/31  - New Trait - Seafoam  - Secondary color (chest color) - green
- Dec/31  - **New Fancy** - PhuZiqaat - Best guess recipe = chartreux + spock + alien + pouty

2018

- Jan/3   - New Trait - Thundergrey - Eye color
- Jan/4   - New Trait - Wonky - Eye type
- Jan/4   - New Trait - Pumpkin - Eye color
- Jan/4   - New Trait - Babypuke  - Eye color
- Jan/6   - New Trait - Grim  - Mouth type
- Jan/6   - New Trait - Grimace  - Mouth type
- Jan/6   - New Trait - Neckbeard  - Mouth type
- Jan/6   - New Trait - Stunned  - Eye type




### Owners / Producers

**Axiom Zen** (web: [axiomzen.co](https://www.axiomzen.co), twitter: [axiomzenteam](https://twitter.com/axiomzenteam), github: [axiomzen](https://github.com/axiomzen))

Founded in 2013. Headquarters in Vancouver, British Columbia (BC), Canada.

> Axiom Zen is a venture studio. We specialize in new venture creation - building products and companies
> at the cutting edge of technology.
>
> Our team of fifty founder-quality engineers, product leads, and designers combine a wealth of experience from fast-growing startups
> and companies.

CEO: Roham Gharegozlou - twitter: [rohamg](https://twitter.com/rohamg), angel: [roham](https://angel.co/roham)

> We recommend curating your kitties collection and enjoying the game [for life],
> as opposed to engaging in [get-rich-quick] speculation. #HODL
>
> -- [Roham Gharegozlou](https://twitter.com/rohamg/status/940979947916623873)



#### Team

_Meet the CryptoKitties Makers_

![](i/cryptokitties-team.png)

(Source: [CryptoKittes For Press / Team](https://www.cryptokitties.co/press))


CryptoKitty Team :

- **Mack Flavelle**  (*) / Product / twitter: [mackflavelle](https://twitter.com/mackflavelle), reddit: [mackflavelle](https://www.reddit.com/u/mackflavelle)
- **Arthur Camara** (*) / Product and Development / twitter: [arthur_camara](https://twitter.com/arthur_camara), reddit: [arthurcamara1](https://www.reddit.com/u/arthurcamara1)
- **Dieter Shirley**  (*)/ Director of Blockchain / twitter: [dete73](https://twitter.com/dete73), reddit: [dete73](https://www.reddit.com/u/dete73), github: [dete](https://github.com/dete)
- **Fabiano Pereira** (*) / Blockchain Contract Scripter / twitter: [FlockonUS](https://twitter.com/FlockonUS), reddit: [flockonus](https://www.reddit.com/u/flockonus)

NEWS.md

![](i/kitties-per-day-2021-02-05.png)

![](i/kittyhelper-players-2021-02-05.png)

> - Does anyone buy anymore? Had this kitties for years. Anyone wanna buy?
> - I have 340 kitties... and I wanna bulk SELL
> - Kitties since 2017. How much does it value right now? Wanna SELL
> - How much should I sell my gen 0 cat for?
> - I stupidly bought these little bastards. Have 6, what's the best thing to do with them - any chance of making any of my lost money back?
> - Pricing help? How much are my kitties worth?
> - I put these kitties up for sale. Can you please tell me if my pricing is right.
> - Can someone help me sell my cats?
> - Are these kittens worth anything?
>
> -- [CryptoKitties](https://www.reddit.com/r/CryptoKitties)

![](i/kitty-sales-per-day-2021-02-05.png)


# CryptoKitties - What's News?

## 2021

### Ethereum Gas Price - All Time High - Transaction Fees


> Birthing fees are pretty high.
> I would like to breed but i saw the transaction free / cost which are at $1000.
>
> I'm trying to buy one, and set the gas price to slow (1hr). Over three hours later, and up to $8.53 on gas & still no cat.
> Is the game doomed?
> Update! The transaction took 6 hrs & 15 min. - but it did finally go through.
>
> Just bought two cheap kitties and tried to breed them. 
> But the birthding fee is 0.1 ETH (~150 USD)... that's crazy.
>
> Yeah, CryptoKitties is unplayable right now.
>
> Yeah, after buying two cats unaware it would cost 0.1 ETH (~150 USD ) to breed them 
> I am a little upset. This is way too expensive, 
> I see absolutely no way to get started for less than a few hundred dollars worth of ETH 
> if you want to breed cats that are worth selling.
>
> Fees are exceptionally high now due to Ethereum's high price and movements.
> The gas fees are tied to Ether (ETH) therefore high prices.
>
> Probably stuck with my cats forever... jeez, I hate to think of how much they cost in dollars with ether (ETH) as expensive as it is now.
>
> How much gas limit should I use to adopt a cat?
> The recommended gas limit in meta mask is absurd and the transaction
> fee would be like $800, how much should i use?
>
> Gas prices [on Ethereum] have now nearly completely stopped the market.
> I know there's plans to revitalize CryptoKitties [on Flow].
>
> People are mostly buying wrapped in bulk to lower gas fees. 
> With the wrapped kitty tokens (WCK, WG0, WVG0) you can buy ten cats and pay $30 fee vs $300 (10 x $30) fees.




### CryptoKitties v2.0 - From Ethereum to Flow (Blockchain) - Build In Public

> On Ethereum, your CryptoKitties live in the storage area associated with the CryptoKitties contract,
> and they have a data field to track who owns them.
>
> On Flow, your CryptoKitties will live in the storage area associated with YOUR account...
>
> -- [Dieter Shirley, Co-Creator CryptoKitties, Creator Flow @ Dapper Labs](https://twitter.com/dete73/status/1301217801101807616)



> Welcome to the early stages of our grand meowgration to Flow.
> We want to enable mainstream adoption of the mascot of blockchain through free to play, new mechanics,
> upgradable smart contracts, easy on-boarding, fewer limitations and restrictions,
> and yes, fully composable, extensible on-chain layer one Kitties with true utility.
> Phew, that was a lot but every bit of it matters.
>
> Our vision is to inspire the world to collect CryptoKitties.
>
> This board will contain ideas, implementations, and designs of all the new features we want to test and potentially bring to Flow.
> Let's put lot of emphasis on the word "test". These features are in various stages of development
> as we work towards building the closed alpha of CryptoKitties on Flow.

![](i/cryptokitties-v2-board.png)

[CryptoKittes on Flow (Blockchain) Board - Build in Public](https://trello.com/b/MVwdMAn7/ck-on-flow-build-in-public)
- [Intrinsic Genes & Nests](https://trello.com/c/89DL79It/25-intrinsic-genes)
  - > Intrinsic genes modify the breeding and genetic properties of CryptoKitties and occupy 4 loci in a Kitty's genome.
    > Types of intrinsic genes:
    >
    > Breeding Properties:
    > - Vigor: modifies breeding time
    > - Fecundity: modifies Catpacity
    >
    > Genetic Properties:
    > - Mutability: modifies both base mutation and devolution probability
    > - Penetrance: modifies rate of recessive emergence (gene swaps within a Locus)
- [Gene Science](https://trello.com/c/ROGqDiMM/20-gene-science)
  - > The Gene Science contract is responsible for the genetic structure of Kitties.
    > Unlike CryptoKitties' math-based mutation pyramid, Gene Science on Flow uses a queryable lookup table to enforce breeding
    > logic for mutating pairs.
    >
    > Every known legacy mutation will be ported over to CryptoKitties on Flow and locked forever.
    > While legacy mutations will remain untouched, we do want your Jaguar to be able to mutate
    > with new CK on Flow genes while retaining its legacy mutation.
    >
    > A gene size of 10 bits will allow for 2^10 potential genes in a single locus.
    > 25 loci make up the CryptoKitties genome with the first 12 being reserved for legacy loci.
    > We've also made changes in the technical implementation of mixgenes and the core Kitty structure to allow backwards compatibility.
- [Breeding with Nests](https://trello.com/c/1LvC0fF0/33-breeding-with-nests)
  - > ![](i/cryptokitties-v2-nest.png)
    >
    > Nests are consumable NFTs required in breeding. To breed a pair of Kitties, you'll need one Nest and available
    > Cat-pacity from a parent. Upon breeding, an Egg NFT is created and placed within your chosen Nest which can be hatched at a later time.
    >
    > Nests have their own genes which can affect the breeding or genetic outcome of your Kitties.
    > Not all of these properties are included in-scope.
    >
    > - Level: Maximum Level of Cat born in this nest
    > - Nestpacity: Number of times the nest can be used before disintegration

mixGenes.rb
#################
#  CryptoKitties geneScience.mixGenes - "magic" sooper-sekret gene mixing operation forumula (ruby edition)
#    to run use:
#      $ ruby ./mixGenes.rb
#
#
#  based on the pseudo-code published by Sean Soria
#    in "CryptoKitties mixGenes Function"
#    see https://medium.com/@sean.soria/cryptokitties-mixgenes-function-69207883fc80
#
#  original call in kitty script:
#   uint256 childGenes = geneScience.mixGenes(matron.genes, sire.genes, matron.cooldownEndBlock - 1);
#
#   see https://etherscan.io/address/0xf97e0a5b616dffc913e72455fde9ea8bbe946a2b#code
#   opscode:
#
#  PUSH1 0x60
#  PUSH1 0x40
#  MSTORE
#  PUSH1 0x04
#  CALLDATASIZE
#  LT
#  PUSH2 0x006c
#  JUMPI
#  PUSH4 0xffffffff
#  PUSH29 0x0100000000000000000000000000000000000000000000000000000000
#  PUSH1 0x00
#  CALLDATALOAD
#  DIV
#  AND
#  PUSH4 0x0d9f5aed
#  ...
#
#  pseudo code:
#
# def mixGenes(mGenes[48], sGenes[48], babyGenes[48]):
#    # PARENT GENE SWAPPING
#  for (i = 0; i < 12; i++):
#    index = 4 * i
#    for (j = 3; j > 0; j--):
#      if random() < 0.25:
#        swap(mGenes, index+j, index+j-1)
#      if random() < 0.25:
#        swap(sGenes, index+j, index+j-1)
#  # BABY GENES
#  for (i = 0; i < 48; i++):
#    mutation = 0
#    # CHECK MUTATION
#    if i % 4 == 0:
#      gene1 = mGene[i]
#      gene2 = sGene[i]
#      if gene1 > gene2:
#        gene1, gene2 = gene2, gene1
#      if (gene2 - gene1) == 1 and iseven(gene1):
#        probability = 0.25
#        if gene1 > 23:
#          probability /= 2
#        if random() < probability:
#          mutation = (gene1 / 2) + 16
#    # GIVE BABY GENES
#    if mutation:
#      baby[i] = mutation
#    else:
#      if random() < 0.5:
#        babyGenes[i] = mGene[i]
#      else:
#        babyGenes[i] = sGene[i]



def mixgenes( mgenes, sgenes )  ## returns babygenes
  ## note: reverse genes strings (in kai) so index 0 is the first number
  ##                                         index 1 is the second number etc.
  mgenes = mgenes.reverse
  sgenes = sgenes.reverse

  babygenes = "?"*48      ## string with 48 question marks (?)

  # PARENT GENE SWAPPING
  12.times do |i| # loop from 0 to 11        # for(i = 0; i < 12; i++)
    puts "parent gene swapping i: #{i}"
    index = 4*i                              #   index = 4 * i
    3.downto(1) do |j| ## loop from 3 to 1   #   for (j = 3; j > 0; j--)
      puts "   j: #{j}"
      if rand(100) < 25                      #     if random() < 0.25:
        mgenes[index+j-1], mgenes[index+j] = #       swap(mGenes, index+j, index+j-1)
        mgenes[index+j],   mgenes[index+j-1]
      end
      if rand(100) < 25                      #     if random() < 0.25:
        sgenes[index+j-1], sgenes[index+j] = #        swap(sGenes, index+j, index+j-1)
        sgenes[index+j],   sgenes[index+j-1]
      end
    end
  end

  # BABY GENES
  48.times do |i| # loop from 0 to 47        #  for (i = 0; i < 48; i++):
    puts "baby genes i: #{i}"
    mutation = nil                           #    mutation = 0
                                             #    # CHECK MUTATION
    if i % 4 == 0                            #    if i % 4 == 0:
      gene1 = Kai::NUMBER[ mgenes[i] ]       #      gene1 = mGene[i]
      gene2 = Kai::NUMBER[ sgenes[i] ]       #      gene2 = sGene[i]
      if gene1 > gene2                       #      if gene1 > gene2:
         gene1, gene2 = gene2, gene1        #        gene1, gene2 = gene2, gene1
      end
      if (gene2 - gene1) == 1 && gene1.even? #     if (gene2 - gene1) == 1 and iseven(gene1):
        probability = 25                     #        probability = 0.25
        if gene1 > 23                        #        if gene1 > 23:
          probability /= 2                   #          probability /= 2
        end
        if rand(100) < probability                   #        if random() < probability:
          mutation = Kai::ALPHABET[ (gene1/2)+16 ]   #          mutation = (gene1 / 2) + 16
        end
      end
    end
    # GIVE BABY GENES
    if mutation                              #    if mutation:
      babygenes[i]=mutation                   #      baby[i] = mutation
    else                                     #    else:
      if rand(100) < 50                      #      if random() < 0.5:
        babygenes[i] = mgenes[i]             #        babyGenes[i] = mGene[i]
      else                                   #      else:
        babygenes[i] = sgenes[i]             #        babyGenes[i] = sGene[i]
      end
    end
  end

  babygenes.reverse   # return bagygenes (reversed back)
end # mixgenes


### let's add base32 / kai helper class

class Base32   ## Base32  (2^5 - 5bits)

  # See https://en.wikipedia.org/wiki/Base58
  ALPHABET = "123456789abcdefghijkmnopqrstuvwx"
  BASE     = ALPHABET.length   ## 32 chars/letters/digits

  # Converts a base10 integer to a base32 string.
  def self.encode( num )
    buf = String.new
    while num >= BASE
      mod = num % BASE
      buf = ALPHABET[mod] + buf
      num = (num - mod)/BASE
    end
    ALPHABET[num] + buf
  end

  NUMBER = {
   '1' => 0,  '2' => 1,  '3' => 2,  '4' => 3,  '5' => 4,  '6' => 5,  '7' => 6,  '8' => 7,
   '9' => 8,  'a' => 9,  'b' => 10, 'c' => 11, 'd' => 12, 'e' => 13, 'f' => 14, 'g' => 15,
   'h' => 16, 'i' => 17, 'j' => 18, 'k' => 19, 'm' => 20, 'n' => 21, 'o' => 22, 'p' => 23,
   'q' => 24, 'r' => 25, 's' => 26, 't' => 27, 'u' => 28, 'v' => 29, 'w' => 30, 'x' => 31
  }

  def self.fmt( kai )
    ## format in groups of four (4) separated by space
    ##  e.g.  ccac7787fa7fafaa16467755f9ee444467667366cccceede
    ##     :  ccac 7787 fa7f afaa 1646 7755 f9ee 4444 6766 7366 cccc eede
    kai.reverse.gsub( /(.{4})/, '\1 ').reverse.strip
  end
end  # class Base32


## add a shortcut (convenience) alias
Kai = Base32
# why kai?
#  in honor of Kai Turner who deciphered the genes - thanks!
#    see https://medium.com/@kaigani/the-cryptokitties-genome-project-on-dominance-inheritance-and-mutation-b73059dcd0a4



#####################
# let's go

mgenes_hex = 0x000063169218f348dc640d171b000208934b5a90189038cb3084624a50f7316c
sgenes_hex = 0x00005a13429085339c6521ef0300011c82438c628cc431a63298e3721f772d29


mgenes = Kai.encode( mgenes_hex )   # convert to 5-bit (base32/kai) notation
p mgenes
# => "ddca578ka4f7949p4d11535kaeea175h846k2243aa9gfdcd"
p Kai.fmt( mgenes )
# => "ddca 578k a4f7 949p 4d11 535k aeea 175h 846k 2243 aa9g fdcd"
p mgenes.size
# => 48

sgenes = Kai.encode( sgenes_hex )
p sgenes
# => "c9am65567ff7b9gg1d1138539f77647577k46784f9gpfcaa"
p Kai.fmt( sgenes )
# => "c9am 6556 7ff7 b9gg 1d11 3853 9f77 6475 77k4 6784 f9gp fcaa"
p sgenes.size
# => 48

babygenes = mixgenes( mgenes, sgenes )
p babygenes
# => "9dca5586aff7b99p1d1133k5aea767h574kk6744aafgffaa"
p Kai.fmt( babygenes )
# => "9dca 5586 aff7 b99p 1d11 33k5 aea7 67h5 74kk 6744 aafg ffaa"
